### PR TITLE
Adding parsing and semantic check support for omp masked

### PR DESCRIPTION
**Pull request by anchuraj**
omp masked directive in OpenMP 5.2 allows to specify code regions which are expected to be executed by thread ids specified by the programmer. Filter clause of the directive allows to specify the thread id. This change adds the parsing support for the directive

### CODE REVIEW

File: flang/lib/Semantics/resolve-directives.cpp
Comment (skatrak) [inline flang/lib/Semantics/resolve-directives.cpp:55]: It appears that we don't really need to store this to check anything, so the `SetMaskedTId` and `GetMaskedTId` functions are not necessary either at this stage. So, If I'm not wrong, these MASKED-related constructs could also just be treated as regular block constructs (no need for special cases in `Pre` and `Post`).

Please do let me know if you don't agree and you had some later uses planned for this.

File: flang/lib/Semantics/resolve-directives.cpp
Comment (skatrak) [inline flang/lib/Semantics/resolve-directives.cpp:1115]: Nit: Unrelated formatting change.

**Review by skatrak — COMMENTED** [review]
Thank you Anchu, this looks mostly good to me. I just have a couple of comments mostly to generally reduce the scope of the changes to only what's necessary.

**Review by mjklemm — CHANGES_REQUESTED** [review]

**Review by kiranchandramohan — CHANGES_REQUESTED** [review]
Thanks for this patch.

Could you add a relevant TODO message in lowering so that any test using masked will not crash. We are trying to stabilize the code by reducing crashes. The behaviour we are trying to get to is that the compiler either accepts the code and executes correctly or provides a TODO message saying it is not implemented yet.

File: flang/lib/Semantics/resolve-directives.cpp
Comment (kiranchandramohan) [inline flang/lib/Semantics/resolve-directives.cpp:1519]:
```suggestion
  // be the primary thread id.
```

File: flang/lib/Semantics/resolve-directives.cpp
Comment (kiranchandramohan) [inline flang/lib/Semantics/resolve-directives.cpp:1508]: Does this work correctly if there is a variable or expression in the filter clause?

File: flang/lib/Parser/openmp-parsers.cpp
Comment (anchuraj) [reply flang/lib/Parser/openmp-parsers.cpp]: Thank you! deleted the formatting changes

File: flang/lib/Parser/openmp-parsers.cpp
Comment (anchuraj) [reply flang/lib/Parser/openmp-parsers.cpp]: Thank you! deleted the formatting changes

**Comment by anchuraj** [comment]
Thank you @skatrak @mjklemm @kiranchandramohan for reviewing the PR.  I have removed the unrelated formatting changes and am working on the other suuggestions.

File: flang/lib/Semantics/resolve-directives.cpp
Comment (mjklemm) [reply flang/lib/Semantics/resolve-directives.cpp:1519]: Maybe even add "the ID of the primary thread (= 0)".

File: flang/lib/Semantics/resolve-directives.cpp
Comment (anchuraj) [reply flang/lib/Semantics/resolve-directives.cpp]: Hi @skatrak , I do agree. I have removed the changes.

Comment (anchuraj) [reply flang/lib/Semantics/resolve-directives.cpp]: Thankyou @mjklemm , the changes are removed as they were incorrect.

File: flang/lib/Semantics/resolve-directives.cpp
Comment (anchuraj) [reply flang/lib/Semantics/resolve-directives.cpp]: I was wrong to treat it as a constant expression. Thanks for pointing this out. I have removed the changes.

File: flang/lib/Semantics/resolve-directives.cpp
Comment (anchuraj) [reply flang/lib/Semantics/resolve-directives.cpp]: The above changes are removed. I will handle these suggestions when I implement lowering of the operation.

File: flang/test/Lower/OpenMP/Todo/masked-directive.f90
Comment (kiranchandramohan) [inline flang/test/Lower/OpenMP/Todo/masked-directive.f90:4]: Please use `%not_todo_cmd` to ensure this works correctly in all kind of builds.

**Review by kiranchandramohan — APPROVED** [review]
LGTM. Thanks for the changes. Please wait for @skatrak and/or @mjklemm.

It will help gracefully erroring out if you can make the OpenMPIRBuilder changes before implementing the Flang lowering.

File: flang/lib/Semantics/resolve-directives.cpp
Comment (skatrak) [inline flang/lib/Semantics/resolve-directives.cpp:1506]: Remove TASKLOOP-related enums from here and `Post(const parser::OpenMPBlockConstruct &)` and instead introduce them in the `Pre` method for `OpenMPLoopConstruct`.

File: flang/test/Lower/OpenMP/Todo/masked-directive.f90
Comment (skatrak) [inline flang/test/Lower/OpenMP/Todo/masked-directive.f90:8]: Nit: Use "!" rather than "//" for Fortran comments.
```suggestion
  ! CHECK: not yet implemented: Unhandled Directive masked
```

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (skatrak) [inline flang/lib/Lower/OpenMP/OpenMP.cpp:2226]: Nit: Add comment with the spec (v4.5, if I'm not wrong, you can check the others below to make sure) section number for "MASKED".

**Review by skatrak — COMMENTED** [review]
Thanks for addressing my feedback, it's looking nearly ready to me. Just some minor comments.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (mjklemm) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:2226]: `masked` is a 5.1 feature

Comment (mjklemm) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:2226]: Oh, if you want the Section number: it's Section 10.5 in OpenMP version 5.2

File: flang/test/Lower/OpenMP/Todo/masked-directive.f90
Comment (mjklemm) [reply flang/test/Lower/OpenMP/Todo/masked-directive.f90:8]: Please fix these comments before merging.

**Comment by mjklemm** [comment]
It might be a good idea to have the compiler emit a warning about the deprecation of the `master` construct and that it should be replaced with `masked`.  Would you agree?  I'm OK with doing that in this PR or a separate PR.

File: flang/include/flang/Semantics/openmp-directive-sets.h
Comment (anchuraj) [reply flang/include/flang/Semantics/openmp-directive-sets.h]: Thank you! Moved the constructs and made the necessary changes

File: flang/test/Lower/OpenMP/Todo/masked-directive.f90
Comment (anchuraj) [reply flang/test/Lower/OpenMP/Todo/masked-directive.f90]: Yeah, thanks. Corrected. Its a copy paste mistake. I noticed a few more examples in the same folder with the same mistake. Will correct those in a separate PR.

File: flang/lib/Semantics/resolve-directives.cpp
Comment (anchuraj) [reply flang/lib/Semantics/resolve-directives.cpp]: Done

File: flang/test/Lower/OpenMP/Todo/masked-directive.f90
Comment (anchuraj) [reply flang/test/Lower/OpenMP/Todo/masked-directive.f90]: Done. Thank you!

**Comment by anchuraj** [comment]
> It might be a good idea to have the compiler emit a warning about the deprecation of the `master` construct and that it should be replaced with `masked`. Would you agree? I'm OK with doing that in this PR or a separate PR.

I feel such a warning would be appropriate after `masked` is supported, since, until then users wont be able to replace. Please let me know your thoughts.

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (anchuraj) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:2226]: I see that in a recent [commit](https://github.com/llvm/llvm-project/pull/90098/files#diff-496a295679ae3c43f8651c944a1bd9dca177ad2b5e4d7121f96938024e292bc1L2217) all section numbers are removed. The section numbers keeps changing unless some one needs to go to the exact spec. So I am more in favor of not adding it. Please let me know if you feel otherwise

Comment (mjklemm) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:2226]: I won't argue for having the section numbers, as they are a moving target (like you rightfully say) and finding the constructs in the OpenMP spec is fairly easy. :-)

**Comment by mjklemm** [comment]
> I am working on end to end support. I will add this as a future PR, if you agree @mjklemm

Works for me!  Thanks!

**Comment by anchuraj** [comment]
> > I am working on end to end support. I will add this as a future PR, if you agree @mjklemm
> 
> Works for me! Thanks!

Thank you! I believe I have addressed the requested changes.  If you are good with the change, can you please approve @mjklemm . I will wait for @skatrak 's approval as well.

**Review by mjklemm — APPROVED** [review]
LGTM.  Thanks!

**Review by skatrak — APPROVED** [review]
Thank you Anchu, LGTM!

**Pull request merged**
